File: java/reflect/src/fan/reflect/TypeExt.java
package fan.reflect;

import java.io.IOException;

import fan.sys.FanType;
import fan.sys.List;
import fan.sys.UnknownSlotErr;
import fanx.fcode.FConst;
import fanx.fcode.FField;
import fanx.fcode.FMethod;
import fanx.fcode.FType;
import fanx.main.Sys;
import fanx.main.Type;

public class TypeExt {
	private static Type typeof;

	public Type typeof() {
		if (typeof == null) {
			typeof = Sys.findType("reflect::TypeExt");
		}
		return typeof;
	}

	public static Pod pod(Type type) {
		return Pod.fromFPod(type.ftype().pod);
	}

	// public static Map params(Type type) {
	// return null;
	// }
	//
	// public static Type parameterize(Type type, Map params) {
	// return null;
	// }

	public static List fields(Type type) {
		java.util.Map<String, Object> slots = getSlots(type);
		List list = List.make(Slot.typeof, slots.size());
		for (java.util.Map.Entry<String, Object> entry : slots.entrySet()) {
			if (entry.getValue() instanceof Field) {
				list.add(entry.getValue());
			}
		}
		return list;
	}

	public static List methods(Type type) {
		java.util.Map<String, Object> slots = getSlots(type);
		List list = List.make(Slot.typeof, slots.size());
		for (java.util.Map.Entry<String, Object> entry : slots.entrySet()) {
			if (entry.getValue() instanceof Method) {
				list.add(entry.getValue());
			}
		}
		return list;
	}

	public static List slots(Type type) {
		java.util.Map<String, Object> slots = getSlots(type);
		List list = List.make(Slot.typeof, slots.size());
		for (java.util.Map.Entry<String, Object> entry : slots.entrySet()) {
			list.add(entry.getValue());
		}
		return list;
	}

	public static Field field(Type type, String name, boolean checked) {
		return (Field) slot(type, name, checked);
	}

	public static Field field(Type type, String name) {
		return field(type, name, true);
	}

	public static Method method(Type type, String name, boolean checked) {
		return (Method) slot(type, name, checked);
	}

	public static Method method(Type type, String name) {
		return method(type, name, true);
	}
	
	private static void mergeSlots(java.util.Map<String, Object> out, Type base) {
		java.util.Map<String, Object> add = getSlots(base);
		for (java.util.Map.Entry<String, Object> entry : add.entrySet()) {
			out.put(entry.getKey(), entry.getValue());
		}
	}

	private static java.util.Map<String, Object> getSlots(Type type) {
		if (type.slots != null)
			return type.slots;
		java.util.Map<String, Object> slots = new java.util.HashMap<String, Object>();
		
		//get inheritance slots
		if (!type.isObj() && !FanType.isMixin(type)) {
			mergeSlots(slots, type.base());
		}
		List mixins = FanType.mixins(type);
		for (int i=0; i<mixins.size(); ++i) {
			Type t = (Type)mixins.get(i);
			mergeSlots(slots, t);
		}

		//get self type slots
		FType ftype = type.ftype();
		if (ftype != null) {
			ftype.load();
			for (FField f : ftype.fields) {
				if (f.isSynthetic()) continue;
				slots.put(f.name, Field.fromFCode(f, type));
			}
			
			for (FMethod f : ftype.methods) {
				if ((f.flags & FConst.Getter) != 0) {
					Field field = (Field)slots.get(f.name);
					field.getter = Method.fromFCode(f, type);
					continue;
				}
				if ((f.flags & FConst.Setter) != 0) {
					Field field = (Field)slots.get(f.name);
					field.setter = Method.fromFCode(f, type);
					continue;
				}
				
				if (f.isSynthetic()) continue;
				slots.put(f.name, Method.fromFCode(f, type));
			}
		}
		
		//link java reflect
		java.lang.reflect.Method[] jmths = type.getJavaClass().getMethods();
		for (java.lang.reflect.Method jmth : jmths) {
			Slot s = (Slot)slots.get(jmth.getName());
			if (s == null) continue;
			if (s instanceof Method) {
				Method mth = (Method)s;
				if (jmth.getParameterCount() < mth.reflect.length) {
					mth.reflect[jmth.getParameterCount()] = jmth;
				}
			}
			else if (s instanceof Field) {
				Field field = (Field)s;
				if (jmth.getReturnType() == void.class)
			        field.setter.reflect = new java.lang.reflect.Method[] { jmth };
			    else
			        field.getter.reflect = new java.lang.reflect.Method[] { jmth };
			}
			jmth.setAccessible(true);
		}
		
		type.slots = slots;
		return type.slots;
	}

	public static Slot slot(Type type, String name, boolean checked) {
		Slot s = (Slot) getSlots(type).get(name);
		if (s == null) {
			if (checked) {
				throw UnknownSlotErr.make(name);
			}
		}
		return s;
	}

	public static Slot slot(Type type, String name) {
		return slot(type, name, true);
	}
}

File: src/std/js/TypeExt.js

fan.std.TypeExt = fan.sys.Obj.$extend(fan.sys.Obj);
fan.std.TypeExt.prototype.$ctor = function() {}
fan.std.TypeExt.prototype.$typeof = function() { return fan.std.TypeExt.$type; }

fan.std.TypeExt.hasFacet = function(self, type)
{
  return self.hasFacet(type);
}

fan.std.TypeExt.facets = function(self)
{
  return self.facets();
}

fan.std.TypeExt.facet = function(self, type, checked)
{
  return self.facet(type, checked);
}

fan.std.TypeExt.slots   = function(self) { return self.slots(); }
fan.std.TypeExt.methods = function(self) { return self.methods(); }
fan.std.TypeExt.fields  = function(self) { return self.fields(); }

fan.std.TypeExt.slot = function(self, name, checked)
{
  return self.slot(name, checked);
}

fan.std.TypeExt.method = function(self, name, checked)
{
  return self.method(name, checked);
}

fan.std.TypeExt.field = function(self, name, checked)
{
  return self.field(name, checked);
}

fan.std.TypeExt.pod = function(self)
{
  return self.pod();
}

fan.std.TypeExt.doc = function(self)
{
  return self.doc();
}

fan.sys.TypeExt.emptyList = function(self)
{
  if (self.$emptyList == null) {
    self.$emptyList = fan.sys.List.make(0, self);
    //.toImmutable();
    self.$emptyList.m_readOnly = true;
    self.$emptyList.m_immutable = true;
  }
  return self.$emptyList;
}

//////////////////////////////////////////////////////////////////////////
// Make
//////////////////////////////////////////////////////////////////////////

fan.std.TypeExt.make = function(self, args)
{
  if (args === undefined) args = null;

  var make = self.method("make", false);
  if (make != null && make.isPublic())
  {
    if (self.isAbstract() && !make.isStatic())
      throw fan.sys.Err.make("Cannot instantiate abstract class: " + self.m_qname);

    var numArgs = args == null ? 0 : args.size();
    var params = make.params();
    if ((numArgs == params.size()) ||
        (numArgs < params.size() && params.get(numArgs).hasDefault()))
      return make.invoke(null, args);
  }

  var defVal = self.slot("defVal", false);
  if (defVal != null && defVal.isPublic())
  {
    if (defVal instanceof fan.std.Field) return defVal.get(null);
    if (defVal instanceof fan.std.Method) return defVal.invoke(null, null);
  }

  throw fan.sys.Err.make("Type missing 'make' or 'defVal' slots: " + self);
}

//////////////////////////////////////////////////////////////////////////
// Slots
//////////////////////////////////////////////////////////////////////////

fan.sys.Type.prototype.slots   = function() { return this.reflect().m_slotList.ro(); }
fan.sys.Type.prototype.methods = function() { return this.reflect().m_methodList.ro(); }
fan.sys.Type.prototype.fields  = function() { return this.reflect().m_fieldList.ro(); }

fan.sys.Type.prototype.slot = function(name, checked)
{
  if (checked === undefined) checked = true;
  var slot = this.reflect().m_slotsByName[name];
  if (slot != null) return slot;
  if (checked) throw fan.sys.UnknownSlotErr.make(this.m_qname + "." + name);
  return null;
}

fan.sys.Type.prototype.method = function(name, checked)
{
  var slot = this.slot(name, checked);
  if (slot == null) return null;
  return fan.sys.ObjUtil.coerce(slot, fan.std.Method.$type);
}

fan.sys.Type.prototype.field = function(name, checked)
{
  var slot = this.slot(name, checked);
  if (slot == null) return null;
  return fan.sys.ObjUtil.coerce(slot, fan.std.Field.$type);
}

// addMethod
fan.sys.Type.prototype.$am = function(name, flags, returns, params, facets)
{
  var r = fan.sys.Type.find(returns);
  var m = new fan.std.Method(this, name, flags, r, params, facets);
  this.m_slotsInfo.push(m);
  return this;
}

// addField
fan.sys.Type.prototype.$af = function(name, flags, of, facets)
{
  var t = fan.sys.Type.find(of);
  var f = new fan.std.Field(this, name, flags, t, facets);
  this.m_slotsInfo.push(f);
  return this;
}

//////////////////////////////////////////////////////////////////////////
// Facets
//////////////////////////////////////////////////////////////////////////

fan.sys.Type.prototype.hasFacet = function(type)
{
  return this.facet(type, false) != null;
}

fan.sys.Type.prototype.facets = function()
{
  if (this.m_inheritedFacets == null) this.loadFacets();
  return this.m_inheritedFacets.list();
}

fan.sys.Type.prototype.facet = function(type, checked)
{
  if (checked === undefined) checked = true;
  if (this.m_inheritedFacets == null) this.loadFacets();
  return this.m_inheritedFacets.get(type, checked);
}

fan.sys.Type.prototype.loadFacets = function()
{
  var f = this.m_myFacets.dup();
  var inheritance = this.inheritance();
  for (var i=0; i<inheritance.size(); ++i)
  {
    var x = inheritance.get(i);
    if (x.m_myFacets) f.inherit(x.m_myFacets);
  }
  this.m_inheritedFacets = f;
}

//////////////////////////////////////////////////////////////////////////
// Util
//////////////////////////////////////////////////////////////////////////

fan.sys.Type.prototype.reflect = function()
{
  if (this.m_slotsByName != null) return this;
  this.doReflect();
  return this;
}

fan.sys.Type.prototype.doReflect = function()
{
  // these are working accumulators used to build the
  // data structures of my defined and inherited slots
  var slots = [];
  var nameToSlot  = {};   // String -> Slot
  var nameToIndex = {};   // String -> Int

  // merge in base class and mixin classes
  if (this.m_mixins) {
    for (var i=0; i<this.m_mixins.size(); i++) this.$mergeType(this.m_mixins.get(i), slots, nameToSlot, nameToIndex);
  }
  this.$mergeType(this.m_base, slots, nameToSlot, nameToIndex);

  // merge in all my slots
  for (var i=0; i<this.m_slotsInfo.length; i++)
  {
    var slot = this.m_slotsInfo[i]
    this.$mergeSlot(slot, slots, nameToSlot, nameToIndex);
  }

  // break out into fields and methods
  var fields  = [];
  var methods = [];
  for (var i=0; i<slots.length; i++)
  {
    var slot = slots[i];
    if (slot instanceof fan.std.Field) fields.push(slot);
    else methods.push(slot);
  }

  // set lists
  this.m_slotList    = fan.sys.List.makeFromJs(fan.std.Slot.$type, slots);
  this.m_fieldList   = fan.sys.List.makeFromJs(fan.std.Field.$type, fields);
  this.m_methodList  = fan.sys.List.makeFromJs(fan.std.Method.$type, methods);
  this.m_slotsByName = nameToSlot;
}

/**
 * Merge the inherit's slots into my slot maps.
 *  slots:       Slot[] by order
 *  nameToSlot:  String name -> Slot
 *  nameToIndex: String name -> Long index of slots
 */
fan.sys.Type.prototype.$mergeType = function(inheritedType, slots, nameToSlot, nameToIndex)
{
  if (inheritedType == null) return;
  var inheritedSlots = inheritedType.reflect().slots();
  for (var i=0; i<inheritedSlots.size(); i++)
    this.$mergeSlot(inheritedSlots.get(i), slots, nameToSlot, nameToIndex);
}

/**
 * Merge the inherited slot into my slot maps.  Assume this slot
 * trumps any previous definition (because we process inheritance
 * and my slots in the right order)
 *  slots:       Slot[] by order
 *  nameToSlot:  String name -> Slot
 *  nameToIndex: String name -> Long index of slots
 */
fan.sys.Type.prototype.$mergeSlot = function(slot, slots, nameToSlot, nameToIndex)
{
  // skip constructors which aren't mine
  if (slot.isCtor() && slot.m_parent != this) return;

  var name = slot.m_name;
  var dup  = nameToIndex[name];
  if (dup != null)
  {
    // if the slot is inherited from Obj, then we can
    // safely ignore it as an override - the dup is most
    // likely already the same Object method inherited from
    // a mixin; but the dup might actually be a more specific
    // override in which case we definitely don't want to
    // override with the sys::Object version
    if (slot.parent() == fan.sys.Obj.$type)
      return;

    // if given the choice between two *inherited* slots where
    // one is concrete and abstract, then choose the concrete one
    var dupSlot = slots[dup];
    if (slot.parent() != this && slot.isAbstract() && !dupSlot.isAbstract())
      return;

// TODO FIXIT: this is not triggering -- possibly due to how we generate
// the type info via compilerJs?
    // check if this is a Getter or Setter, in which case the Field
    // trumps and we need to cache the method on the Field
    // Note: this works because we assume the compiler always generates
    // the field before the getter and setter in fcode
    if ((slot.m_flags & (fan.sys.FConst.Getter | fan.sys.FConst.Setter)) != 0)
    {
      var field = slots[dup];
      if ((slot.m_flags & fan.sys.FConst.Getter) != 0)
        field.m_getter = slot;
      else
        field.m_setter = slot;
      return;
    }

    nameToSlot[name] = slot;
    slots[dup] = slot;
  }
  else
  {
    nameToSlot[name] = slot;
    slots.push(slot);
    nameToIndex[name] = slots.length-1;
  }
}


